### DISCUSSION TITLE
Better support for a `definition` syntax

### DISCUSSION THREAD

**Discussion by xade93** [zim-desktop-wiki/zim-desktop-wiki#1706] [Ideas]
## Is your feature request related to a problem? Please describe.
More than often in writing wiki-style texts (especially in math and science fields), definitions about a word / symbol need to be made. Some definitions appear in a recurrent fashion. It would be cool if zim recognizes this pattern, and maintains the creation and later reference of definitions.

## But why not create new page / create new subtitle to accommodate definitions?
 In reality, many definitions can be written in less than one line or so. It would be impractical to create huge title for every new definition introduced. Moreover, as far as I know, current zim has no support for referring to definitions made by a subtitle. 

## Describe the solution you'd like
* Assign a special shortcut for making definitions. Invoking of this shortcut when text is selected automatically format the text into definition format (bold and italic maybe ? but we need to differentiate that and normal formatting so that user know it's a definition) and add the selected text into indexing of definitions.
* An indexing of definitions is maintained for each level of wiki node, like those appear at appendix of books. This can either be calculated on-demand or created and updated by each modification to text file.
* clicking on reference to previous definitions (possibly with ctrl key hold) jumps to the original definition (exact line). This can be queried from index.

* Another shortcut key may be assigned to popup a auto-completion list when we wish to refer to previous definitions.The list should be sorted by order of modify / last use time of definitions.
* Modifying a definition's name shall inquiry user if we need to update the back-links as well.

TLDR: tags, but without the \@ symbol and with indexing & link support

## Trivia
I have been trying various notebook programs for so long. From evernote, wiznote to keepnote, simplenote, pen-paper, obsidian, e-ink tablets ... But Zim Wiki is what settles me down. It has been the single piece of the greatest software I have ever encountered in my 7 year of linux experience. I have been using zim for a year now. Thank you all developers for creating this amazing piece of software. 

Edit: Formatting & Grammar

**Comment by fdt-dev** [comment]
I second your proposal. I'd like to see Zim Wiki support footnotes as well as a way to integrate the contents of any other page into the current page -- like a link but rather displaying the text verbatim. The way TiddlyWiki ([https://tiddlywiki.com/]) can do that.
I agree: Zim Wiki is the single most useful software that I use on a daily basis, and have done so for the last 8 years. Thank you @jaap-karssenberg !

**Reply by introt** [reply]
I've collected some issues related to TiddlyWiki-style transclusion here: https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/1590#issuecomment-922453163

**Comment by jaap-karssenberg** [comment]
The latest version of zim does haev "anchors", so you can include a reference id that other pages can link to.

I think for your idea, the plugin would not need any special syntax, it just can use achor id's and keep track of them.

* Plugin can define a shortcut that kind of acts like a macro and inserts ``**someconcept**: definition of the concept {{id: someconcept}}``
* Next you can link back to the definition using ``[[#someconcept]]`` (just typing ``#someconcept`` will create the link)
* Probably you will also need some function to change the definition (change the name and the id and all uses of the id)
* The plugin could also have an extension for the indexer and keep an index table of all definitions

Alternatively, the definition could be implemented as a special object with it's own widget. Allows for stricter control on e.g. free-form changes of the content. Which might be either a pro or a con. Objects can have reference id's as well, so that mechanism can still be used.

With respect to linking id's, for now they are only supported per page. You can have a special page with definitions. Or we could think of making the linking smarter and also look for anchor id's in other pages if no direct target is specified.

**Comment by supertoothy** [comment]
Linking to definitions per page using link anchors worked really well. I edited style config to make heading 5 small and unobtrusive. Typing # before the word and jumping to its definition works really smoothly now.
